generator/index.ts

```diff
@@ -1,308 +1,3 @@
-}
-
-function generateFacadeFiles() {
-  const schemaLocation: string = process.env.SCHEMA || "";
-  const jujuVersion: string = process.env.JUJU_VERSION || "";
-  const jujuGitSHA: string = process.env.JUJU_GIT_SHA || "";
-  let schema: Array<Facade>;
-  try {
-    const schemaData: string = readFileSync(resolve(schemaLocation), {
-      encoding: "utf8",
-    });
-    schema = JSON.parse(schemaData);
-  } catch (e) {
-    console.error("Unable to parse schema", e);
-    process.exit(1);
-  }
-
-  schema.forEach(async (facade) => {
-    const facadeTemplateData: FacadeTemplate = {
-      name: facade.Name,
-      version: facade.Version,
-      methods: generateMethods(facade.Schema.properties),
-      interfaces: generateInterfaces(facade.Schema.definitions),
-      availableTo: facade.AvailableTo,
-      docBlock: facade.Description,
-      jujuVersion,
-      jujuGitSHA,
-    };
-
-    generateFile(facadeTemplateData);
-  });
-}
-
-/**
-  Generate the list of methods available for the facade. While the API may
-  expose methods, the actual data sent over the wire is an RPC call.
-*/
-function generateMethods(methods: SchemaMethods): FacadeMethod[] {
-  const facadeMethods: FacadeMethod[] = Object.entries(methods).map(
-    (method) => {
-      const generatedMethod: FacadeMethod = {
-        name: method[0],
-        params: extractType(method[1], "Params") || "any",
-        result: extractType(method[1], "Result") || "any",
-        docBlock: method[1].description,
-      };
-      return generatedMethod;
-    }
-  );
-  return facadeMethods;
-}
-
-function generateInterfaces(definitions?: SchemaDefinitions): InterfaceData[] {
-  if (!definitions) {
-    return [];
-  }
-  const interfaces = Object.entries(definitions).map(generateInterface);
-  interfaces.push(
-    generateInterface([
-      "AdditionalProperties",
-      { properties: { "[key: string]": { type: "any" } }, type: "object" },
-    ])
-  );
-  return interfaces;
-}
-
-function generateInterface(
-  definition: [string, SchemaDefinition]
-): InterfaceData {
-  let types: InterfaceType[];
-  if (definition[1].properties) {
-    types = generateTypes(
-      definition[1].properties,
-      definition[1].required || []
-    );
-  } else {
-    types = [
-      {
-        name: "[key: string]",
-        type: "AdditionalProperties",
-        // This isn't actually required but we don't want to add the ? for
-        // optional with this key type.
-        required: true,
-      },
-    ];
-  }
-  return {
-    name: definition[0],
-    types,
-  };
-}
-
-function generateTypes(
-  properties: DefinitionProperties,
-  required: string[]
-): InterfaceType[] {
-  function extractType(values: JSONSchemaType): string {
-    if (values.type) {
-      if (
-        values.patternProperties ||
-        (values.additionalProperties && values.type === "object")
-      ) {
-        // There are additional unknown properties defined.
-        return "AdditionalProperties";
-      }
```

Comment:
This is where we were losing a lot of types. Instead of using the types provided by `patternProperties` this was replacing them with `AdditionalProperties` which is an alias for "any object".